# 4.5: Data Consistency Checks

## This script contain the following:
### 1. Checking for mixed data types
### 2. Checking for missing values
### 3. Checking for duplicates
### 4. Data consistency on orders dataset
### 5. Exporting cleaned dataset

In [22]:
#importing libraries

In [2]:
import pandas as pd
import numpy as np
import os

In [10]:
#importing df_prods

In [4]:
path = r'C:/Users/LocalAdmin/Instacart Basket Analysis'

In [6]:
vars_list = ['order_id', 'user_id', 'order_number', 'order_dow', 'order_hour_of_day', 'days_since_prior_order']

In [8]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [23]:
#Checking the imported df_prods

In [26]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [29]:
#importing df_ords

In [10]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [35]:
#checking the imported df_ords

In [38]:
df_ords.head ()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

### 1. Checking for mixed data type

In [48]:
# create a dataframe

In [51]:
df_test = pd.DataFrame()

In [54]:
# create a mixed type column

In [57]:
df_test['mix']=['a','b',1, True]

In [60]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [63]:
#Checking if a dataframe contains mixed data types

In [66]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_14220\1786467950.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [71]:
# dataframe is mixed. Proceed to fix the this

In [74]:
df_test['mix'] = df_test['mix'].astype('str')

### 2. Checking for Missing values

In [80]:
#Finding missing values in df_prods

In [12]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# creating a dataframe of missing values only

In [14]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [95]:
#amputation only works for numeric values. String variables might reuire to be removed from the dataset.

In [18]:
df_prods.shape

(49693, 5)

In [20]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [22]:
df_prods_clean.shape

(49677, 5)

reduction of rows from 49693 to 49677 is because missing values were removed.

### 3. Checking for Duplicates

In [113]:
#finding duplicates

In [24]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [26]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [122]:
#addressing/deleting duplicates

In [28]:
df_prods_clean.shape

(49677, 5)

In [128]:
#dropping duplicates

In [30]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [32]:
df_prods_clean_no_dups.shape

(49672, 5)

### 4. Data consistency checks on orders dataset

In [137]:
#Question 2: describe df_ords and interpret the output

In [140]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

1.The column titled 'Unnamed:0' has no meaning as it is compilled out of indices that locate different rows of the dataframe,
2.Columns order_id and user_id cannot have summary stats as they are meent to identify orders and users,
3. The 'order_number' too isnt just a value used to identify different order thus there is no need of describing them,
4. The remaining columns are okay as their values do not breach the reasonable counts of days in a week, hours in a day and the days since the last order,
5. First column must be dropped whilst order_id, user_id and order_numner are turned into string variables so that they do not feature in the summary stats once the df.describe function is executed.

In [145]:
#Question 3. Check for mixed-type data in your df_ords dataframe.

In [148]:
print('order_id',':',pd.api.types.infer_dtype(df_ords['order_id']))

order_id : integer


In [152]:
print('user_id',':',pd.api.types.infer_dtype(df_ords['user_id']))

user_id : integer


In [155]:
print('order_number',':',pd.api.types.infer_dtype(df_ords['order_number']))

order_number : integer


In [158]:
print('orders_day_of_week',':',pd.api.types.infer_dtype(df_ords['orders_day_of_week']))

orders_day_of_week : integer


In [161]:
print('order_hour_of_day',':',pd.api.types.infer_dtype(df_ords['order_hour_of_day']))

order_hour_of_day : integer


In [164]:
print('days_since_prior_order',':',pd.api.types.infer_dtype(df_ords['days_since_prior_order']))

days_since_prior_order : floating


In [167]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


All the columns have one form of data type and there is no need for changing it. I therefore skip question 4.


In [172]:
#question 5: Run a check for missing values in your df_ords dataframe

In [34]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

Only the column 'days_since_prior_order' has missing values (206209). 

In [182]:
#Question 6: Address the missing values using an appropriate method.

In [36]:
df_ords['days_since_prior_order'].fillna(3, inplace = True)

C:\Users\LocalAdmin\AppData\Local\Temp\ipykernel_5584\2005944350.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(3, inplace = True)


In [38]:
df_ords[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     3.0  
11                      11                     3.0  
26                      14                     3.0  
39                      11                     3.0  
45                      12                     3.0  
...                    ...                     ...  
3420930                 12                     3.0  
3420934                 18                     3.0  
3421002                 19                     3.0  
3421019                 15                     3.0  
3421069                 11                     3.0  

[206209 rows x 7 columns]

I used the mean of 'days_since_prior_order', that is approximately 3 days. I used this to replace all missing values where order_number is 1 sicne it is the one with the missing values.

In [194]:
#Question 7. Run a check for duplicate values in your df_ords data.

In [40]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [42]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [204]:
df_ords_dups.head()

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [44]:
#Checking the shape of df_ords after imputation
df_ords.shape

(3421083, 7)

#the results shows that there are no duplicates in the df_ords dataframe so number 8 is skipped.

### 5. Exportimg cleaned dataset

In [212]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_verified.csv'))

In [215]:
df_prods.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_verified.csv'))